url3aa01d34ec_katsuster_ememu_tgz-pJ8-net_katsuster_ememu_ui_MainJ8/src/net/katsuster/ememu/ui/VirtualTerminal.java

```diff
@@ -0,0 +1,386 @@
+        }
+        if ((e.getModifiersEx() & KeyEvent.CTRL_DOWN_MASK) == KeyEvent.CTRL_DOWN_MASK) {
+            ctrl = true;
+        }
+        if ((e.getModifiersEx() & KeyEvent.ALT_DOWN_MASK) == KeyEvent.ALT_DOWN_MASK) {
+            alt = true;
+        }
+
+        if (!ctrl && !alt) {
+            switch (keycode) {
+            case KeyEvent.VK_UP:
+                datArray = new byte[]{'\033', '[', 'A'};
+                validMulti = true;
+                break;
+            case KeyEvent.VK_DOWN:
+                datArray = new byte[]{'\033', '[', 'B'};
+                validMulti = true;
+                break;
+            case KeyEvent.VK_LEFT:
+                datArray = new byte[]{'\033', '[', 'D'};
+                validMulti = true;
+                break;
+            case KeyEvent.VK_RIGHT:
+                datArray = new byte[]{'\033', '[', 'C'};
+                validMulti = true;
+                break;
+            default:
+                dat = keychar;
+                if (keychar == 0xffff) {
+                    valid = false;
+                }
+                break;
+            }
+        } else if (!shift && ctrl && !alt) {
+            //Ctrl + @, A...Z, [, \, ], ^, _
+            if (KeyEvent.VK_A <= keycode && keycode <= KeyEvent.VK_Z) {
+                dat = 0x01 + keycode - KeyEvent.VK_A;
+            } else {
+                switch (keycode) {
+                case KeyEvent.VK_AT:
+                    dat = 0x00;
+                    break;
+                case KeyEvent.VK_OPEN_BRACKET:
+                case KeyEvent.VK_BACK_SLASH:
+                case KeyEvent.VK_CLOSE_BRACKET:
+                    dat = 0x1b + keycode - KeyEvent.VK_OPEN_BRACKET;
+                    break;
+                case KeyEvent.VK_CIRCUMFLEX:
+                    dat = 0x1e;
+                    break;
+                default:
+                    valid = false;
+                    break;
+                }
+            }
+        } else {
+            dat = keychar;
+            if (keychar == 0xffff) {
+                valid = false;
+            }
+        }
+
+        try {
+            if (validMulti) {
+                inPout.write(datArray);
+            } else if (valid) {
+                inPout.write(dat);
+            }
+        } catch (IOException ex) {
+            //ignored
+        }
+    }
+
+    @Override
+    public void keyReleased(KeyEvent e) {
+        //do nothing
+    }
+
+    /**
+     * outPout に出力された文字を画面に印字するクラスです。
+     */
+    private class TextDrainer implements Runnable {
+        public TextDrainer() {
+            //do nothing
+        }
+
+        @Override
+        public void run() {
+            try {
+                int i = 0;
+                while (!shouldHalt()) {
+                    vt.layoutChars(outPin);
+
+                    try {
+                        SwingUtilities.invokeAndWait(new Painter(vt));
+                    } catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)